cmd/meshregistrator/service_registration.go

```diff
@@ -0,0 +1,59 @@
+package main
+
+import (
+	"hash/fnv"
+	"strconv"
+	"sync"
+)
+
+type ServiceRegistrationsList struct {
+	sync.Mutex
+	IsDirty bool
```

Comment:
AFAIR bare `bool` isn't atomic in Golang, there is no guaranty that its changes will be visible in other executable units.

Reply:
IsDirty is only written wrapped in Lock/Unlock, is that still problematic? Stale value during reads shouldn't be an issue, might cause unnecessary updates to cloudmap, but w/e.

Reply:
It could be an issue.

An execution unit doesn't know anything about other users of the variable, so it could assume that it's the only user of the variable.  Assuming that, it could load the variable value into a CPU register, and then never read from or write to the variable's original memory location.  As a result, execution units will use their own version of variable stored in their own copies of registers, and e.g. one of them will never see changes made by another one.
Lock/Unlock doesn't do anything with that issue, they can prevent concurrent changes of the same memory location and partial reads, but cannot prevent caching in registers and cannot enforce reading/writing from/to memory.

I'm not 100% sure that all of that is actually happen in Golang, but short googling of something like "golang atomic bool" shows both special `atomic` functions for `bool` and some discussions about that.

Reply:
intresting.. i tried reproducing this behavior and failed: https://play.golang.org/p/X3i4glvkpQF, it's sending an int around a channel loop between few goroutines and randomly increases the int and flips the value of bool such that bool always matches int parity. if any goroutine was caching the value of bool - we'd see it not matching the parity of an int, but so far it didn't happen. so maybe not an issue in golang?

also, i can't imagine the worst case being anything worse than missing an update or two, which would mean up to 10 seconds stale data in cloudmap - i guess we can live with that?

Reply:
Here is an another one example of concurrent reading and writing of `bool`: https://news.ycombinator.com/item?id=17350776
Reading isn't protected by mutex in this code in the PR:
```
if !pods.IsDirty && !local.IsDirty && !aws.IsDirty {
```
Anyway, I'm really not ready to invest so much energy into this discussion. :)
If you think that this code is fine, let's move on.

Reply:
i've replaced the whole thing with contexts and channels which is more idiomatic in go, thanks for your input!